### PR TITLE
NuMI Flux Sysematics Update

### PR DESCRIPTION
**Pull request by woodtp**
This PR tied closely with [sbndata PR#4 Update NuMI flux file to latest version](https://github.com/SBNSoftware/sbndata/pull/4).

The updates here are necessary to access data correctly from the flux file in the above PR, as the internal structure of the file has undergone several revisions.

Specific Updates:
- The file name referencing the file from `sbndata` that needs to be accessed has been moved from a local variable of `NuMIFluxSyst::Shift` to a private member variable of the `NuMIFluxSyst` class as a whole. My opinion is that it will be easier to locate and update down the line, this way. Although, I'm not a fan of having the filename hard-coded into the class, as it has to be updated separately in the new `NuMIPpfxWeight` class, but I don't have a better solution at the moment.
- @brucehoward-physics wrote a `NuMIPpfxWeight` class and corresponding `Var` to facilitate extracting the NuMI flux weights, so I've incorporated that into this PR. Just as a simple test that it's working as expected, I've pulled a small distribution of flux weights:

image

Other modifications:
- `attenuation` has been renamed to `att`
- `beam_divergence` --> `beam_div`
- `horn1_position_xy` separated into `horn1_x` and `horn1_y`
- `horn_current` --> `horn_current_plus` (the -2kA sample found to be consistent with statistical fluctuations, so we dropped it)
- `horn2_position_xy` removed
- `beam_spot_size` --> `beam_spot`
- `beam_shift_minus_y` and `beam_shift_plus_y` split from `beam_shift_xy` (x-shift samples dropped; consistent with stat. flucts.)
- `B_field` removed

Again to check that this update still works as expected, below are a selection of true neutrino energy spectra varied according to the various systematics.

numi_beam_div
numi_beam_shift_minus_y
numi_beam_shift_plus_y
numi_beam_shift_x
numi_beam_spot
numi_horn1_x
numi_horn1_y
numi_horn_current_plus
numi_pc_0
numi_pc_1
numi_pc_2
numi_pc_3
numi_pc_4
numi_water_layer

### CODE REVIEW

File: sbnana/CAFAna/Systs/NuMIFluxSysts.cxx [hunk 1]
```diff
@@ -84,13 +85,15 @@ namespace ana
...
+    // BH -- attempt to not change any weight if the value is inf or nan, just continue with weight as-is...
+    if( std::isinf(h->GetBinContent(bin)) || std::isnan(h->GetBinContent(bin)) ) weight *= 1;
+    else weight *= 1 + sigma*h->GetBinContent(bin);
   }
 
   //----------------------------------------------------------------------
   const NuMIFluxSyst* GetNuMIFluxSyst(const std::string& dir,
-                                      const std::string& prefix,
-                                      const std::string& name)
+                                                              const std::string& prefix,
```
Comment (brucehoward-physics) [inline sbnana/CAFAna/Systs/NuMIFluxSysts.cxx:95]: Looks like the alignment got offset

File: sbnana/CAFAna/Systs/NuMIFluxSysts.cxx [hunk 2]
```diff
@@ -84,13 +85,15 @@ namespace ana
 
     if(bin == 0 || bin == h->GetNbinsX()+1) return;
 
-    weight *= 1 + sigma*h->GetBinContent(bin);
+    // BH -- attempt to not change any weight if the value is inf or nan, just continue with weight as-is...
+    if( std::isinf(h->GetBinContent(bin)) || std::isnan(h->GetBinContent(bin)) ) weight *= 1;
+    else weight *= 1 + sigma*h->GetBinContent(bin);
```
Comment (brucehoward-physics) [inline sbnana/CAFAna/Systs/NuMIFluxSysts.cxx:90]: I guess this is something I added when I was tinkering with these files in my working area? Was I right to add some protection here: are there potential infs/nans to worry about in the weights?

File: sbnana/CAFAna/Systs/NuMIFluxSysts.h [hunk 3]
```diff
@@ -11,6 +13,8 @@ namespace ana
 
   class NuMIFluxSyst : public ISyst
   {
+    static constexpr std::string_view fluxFileName = "2023-07-06_out_450.37_7991.98_79512.66_QEL11.root";
```
Comment (brucehoward-physics) [inline sbnana/CAFAna/Systs/NuMIFluxSysts.h:16]: Is there a potential use case here where we could actually allow it to declare a location on construction and make the name and location a combined data member which defaults to this file? Maybe not? But asking...

**Comment by brucehoward-physics** [comment]
Thanks for this PR Tony -- It looks like I'm the only person tagged as a reviewer and since some of it is stuff I helped put together, I guess I'm thinking we should add another reviewer 😅 

@PetrilloAtWork is this one for you or... do you have someone else in mind? Maybe @jedori0228?

**Comment by jedori0228** [comment]
Hi, I merged this PR and checked `NuMIFluxSysts` and `NuMIFlux` work as expected. Few validation plots added:

beam_spot.pdf

pc_1.pdf

PPFXW.pdf

**Comment by apwood-physics** [comment]
I've made another commit, which changes the following:

1. New feature to the `NuMIFluxSyst` class allowing the user to specify a path to the fluxfile. It will default to the one defined by the `SBNDATA_DIR` envvar.
2. Additional systematic has been added to account for differences between the older beam geometry and the newer 1 MW upgrade to NuMI.
3. The statistical uncertainty on the flux prediction was omitted in previous versions, so I've added a few lines to load it.
4. Some minor changes to which histograms are loaded for beam focusing systematics. I realized that these were reading in the fractional uncertainties instead of the systematic shifts, which would cause an improper accounting for correlations between bins.

**Review by jedori0228 — APPROVED** [review]
Checked with the new flux file and codes, and we see `beam_power` and `stat` uncertainties properly. Looks good to me.

**Pull request merged**